text/0000-proc-macros.md

```diff
@@ -0,0 +1,417 @@
+extensions, aka compiler plugins). This RFC specifies syntax for the definition
+of procedural macros, a high-level view of their implementation in the compiler,
+and outlines how they interact with the compilation process.
+
+This RFC specifies the architecture of the procedural macro system. It relies on
+[RFC 1561](https://github.com/rust-lang/rfcs/pull/1561) which specifies the
+naming and modularisation of macros. It leaves many of the details for further
+RFCs, in particular the details of the APIs available to macro authors
+(tentatively called `libmacro`). See this [blog post](http://ncameron.org/blog/libmacro/)
+for some ideas of how that might look.
+
+At the highest level, macros are defined by implementing functions marked with
+a `#[macro]` attribute. Macros operate on a list of tokens provided by the
+compiler and return a list of tokens that the macro use is replaced by. We
+provide low-level facilities for operating on these tokens. Higher level
+facilities (e.g., for parsing tokens to an AST) should exist as library crates.
+
+
+# Motivation
+[motivation]: #motivation
+
+Procedural macros have long been a part of Rust and have been used for diverse
+and interesting purposes, for example [compile-time regexes](https://github.com/rust-lang-nursery/regex),
+[serialisation](https://github.com/serde-rs/serde), and
+[design by contract](https://github.com/nrc/libhoare). They allow the ultimate
+flexibility in syntactic abstraction, and offer possibilities for efficiently
+using Rust in novel ways.
+
+Procedural macros are currently unstable and are awkward to define. We would
+like to remedy this by implementing a new, simpler system for procedural macros,
+and for this new system to be on the usual path to stabilisation.
+
+One major problem with the current system is that since it is based on ASTs, if
+we change the Rust language (even in a backwards compatible way) we can easily
+break procedural macros. Therefore, offering the usual backwards compatibility
+guarantees to procedural macros, would inhibit our ability to evolve the
+language. By switching to a token-based (rather than AST- based) system, we hope
+to avoid this problem.
+
+# Detailed design
+[design]: #detailed-design
+
+There are two kinds of procedural macro: function-like and macro-like. These two
+kinds exist today, and other than naming (see
+[RFC 1561](https://github.com/rust-lang/rfcs/pull/1561)) the syntax for using
+these macros remains unchanged. If the macro is called `foo`, then a function-
+like macro is used with syntax `foo!(...)`, and an attribute-like macro with
+`#[foo(...)] ...`. Macros may be used in the same places as `macro_rules` macros
+and this remains unchanged.
+
+To define a procedural macro, the programmer must write a function with a
+specific signature and attribute. Where `foo` is the name of a function-like
+macro:
+
+```
+#[macro]
+pub fn foo(TokenStream, &mut MacroContext) -> TokenStream;
```

Comment:
Alternative: use the `macro` keyword (it was reserved in 1.0). It would make it feel more part of the language, in my opinion.

```rust
pub macro foo(TokenStream, &mut MacroContext) -> TokenStream;
```